File: sample/pascalast/bench/ackermann.pas
program AckBench;
{ Ackermann — deep recursion.  ack(3, 10) = 8189. }
function ack(m, n: integer): integer;
begin
  if m = 0 then ack := n + 1
  else if n = 0 then ack := ack(m - 1, 1)
  else ack := ack(m - 1, ack(m, n - 1))
end;
begin
  writeln(ack(3, 10))
end.

File: sample/astr/test/ackermann.r
# expect: 29
# Ackermann's function — exercises 2-arg recursion
ack <- function(m, n) {
    if (m == 0) {
        n + 1
    } else if (n == 0) {
        ack(m - 1, 1)
    } else {
        ack(m - 1, ack(m, n - 1))
    }
}
print(ack(3, 2))
